modules/navier_stokes/src/actions/NSFVAction.C

```diff
@@ -0,0 +1,2023 @@
+                            params);
+    }
+  }
+
+  const std::string ml_kernel_type = "WallDistanceMixingLengthAux";
+  InputParameters ml_params = _factory.getValidParams(ml_kernel_type);
+  ml_params.set<std::vector<SubdomainName>>("block") = _blocks;
+  ml_params.set<AuxVariableName>("variable") = NS::mixing_length;
+  ml_params.set<std::vector<BoundaryName>>("walls") =
+      getParam<std::vector<BoundaryName>>("mixing_length_walls");
+  if (isParamValid("mixing_length_aux_execute_on"))
+    ml_params.set<ExecFlagEnum>("execute_on") =
+        getParam<ExecFlagEnum>("mixing_length_aux_execute_on");
+  ml_params.set<Real>("von_karman_const") = getParam<Real>("von_karman_const");
+  ml_params.set<Real>("von_karman_const_0") = getParam<Real>("von_karman_const_0");
+  ml_params.set<Real>("delta") = getParam<Real>("mixing_length_delta");
+
+  _problem->addAuxKernel(ml_kernel_type, "mixing_length_aux ", ml_params);
+}
+
+void
+NSFVAction::addINSMomentumPressureKernels()
+{
+  if (_porous_medium_treatment)
+  {
+    const std::string press_kernel_type = "PINSFVMomentumPressure";
+    InputParameters params = _factory.getValidParams(press_kernel_type);
+    params.set<std::vector<SubdomainName>>("block") = _blocks;
+    params.set<UserObjectName>("rhie_chow_user_object") = "pins_rhie_chow_interpolator";
+    params.set<MooseFunctorName>(NS::porosity) = _porosity_name;
+    params.set<CoupledName>("pressure") = {NS::pressure};
+
+    for (unsigned int d = 0; d < _dim; ++d)
+    {
+      params.set<NonlinearVariableName>("variable") = NS::superficial_velocity_vector[d];
+      params.set<MooseEnum>("momentum_component") = NS::directions[d];
+
+      _problem->addFVKernel(
+          press_kernel_type, "pins_momentum_" + NS::directions[d] + "_pressure", params);
+    }
+  }
+  else
+  {
+    const std::string press_kernel_type = "INSFVMomentumPressure";
+    InputParameters params = _factory.getValidParams(press_kernel_type);
+    params.set<std::vector<SubdomainName>>("block") = _blocks;
+    params.set<UserObjectName>("rhie_chow_user_object") = "ins_rhie_chow_interpolator";
+    params.set<CoupledName>("pressure") = {NS::pressure};
+
+    for (unsigned int d = 0; d < _dim; ++d)
+    {
+      params.set<NonlinearVariableName>("variable") = NS::velocity_vector[d];
+      params.set<MooseEnum>("momentum_component") = NS::directions[d];
+
+      _problem->addFVKernel(
+          press_kernel_type, "ins_momentum_" + NS::directions[d] + "_pressure", params);
+    }
+  }
+}
+
+void
+NSFVAction::addINSMomentumGravityKernels()
+{
+  if (isParamValid("gravity"))
+  {
+    if (_porous_medium_treatment)
+    {
+      const std::string grav_kernel_type = "PINSFVMomentumGravity";
+      InputParameters params = _factory.getValidParams(grav_kernel_type);
+      params.set<std::vector<SubdomainName>>("block") = _blocks;
+      params.set<UserObjectName>("rhie_chow_user_object") = "pins_rhie_chow_interpolator";
```

Comment:
umm @lindsayad does everything still need to be a residual object?
Is it helping for the PSPG investigation?

Reply:
I'll just leave these guys here.